pkg/internal/controller/controller.go

```diff
@@ -144,6 +166,42 @@ func (c *Controller[request]) NeedLeaderElection() bool {
 	return *c.LeaderElected
 }
 
+// Warmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) Warmup(ctx context.Context) error {
+	if c.NeedWarmup == nil || !*c.NeedWarmup {
+		return nil
+	}
+	return c.startEventSources(ctx)
+}
+
+// WaitForWarmupComplete returns true if warmup has completed without error, and false if there was
+// an error during warmup. If context is cancelled, it returns true.
+func (c *Controller[request]) WaitForWarmupComplete(ctx context.Context) bool {
+	if c.NeedWarmup == nil || !*c.NeedWarmup {
+		return true
+	}
+	ticker := time.NewTicker(syncedPollPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return true
+		case <-ticker.C:
+			didFinishSync, ok := c.didEventSourcesFinishSyncSuccessfully.Load().(*bool)
```

Comment:
Can we make `didEventSourcesFinishSyncSuccessfully` a channel which we close after having done that so we don't have to poll here?